torch/csrc/jit/codegen/cuda/expr_simplifier.cpp

```diff
@@ -0,0 +1,430 @@
+      addInput(v);
+    }
+  }
+
+  NVFUSER_DECLARE_CLONE_AND_CREATE
+
+  virtual const char* getOpString() const override {
+    return "FlattenedAssocCommOp";
+  }
+
+  DataType dtype() const {
+    return *output(0)->getDataType();
+  }
+
+  BinaryOpType getOpType() const {
+    return attribute(0)->as<Attribute<BinaryOpType>>()->value;
+  }
+
+  // Constant terms are stored and handled separately from other inputs. For
+  // example, if I have a + b + 3 + c + 5, then the constant term will be 8, and
+  // a, b, c are all inputs.
+  bool hasConstantTerm() const {
+    return attributes().size() == 2;
+  }
+
+  Val* getConstantTerm() const {
+    return attributeVal(1);
+  }
+
+  template <typename T>
+  T getConstantTermValue() const {
+    return *getConstantTerm()->as<Scalar<T>>()->value();
+  }
+
+  // Get a vector of inputs, sorted as the order given by `variables`. Note that
+  // the sorting key is the rightmost variable that an input depends on. For
+  // example, if I have two inputs.
+  // v1 = a * c
+  // v2 = b
+  // and variables is [a, b, c], then v2 < v1 because the rightmost depending
+  // variable of v2 is b, and the rightmost depending variable of v1 is c,
+  // and b < c. So in this example, this function will return [v2, v1].
+  // Tensors are always considered as variables and they are always considered
+  // as the rightmost.
+  std::vector<Val*> sortedInputs(const std::list<ValInfo>& variables) {
+    std::unordered_set<Val*> variables_set;
+    variables_set.reserve(variables.size());
+    for (auto v : variables) {
+      variables_set.emplace(v.variable);
+    }
+    std::vector<Val*> sorted_inputs(inputs().begin(), inputs().end());
+    auto compare = [&](Val* v1, Val* v2) {
+      bool v1_lefter_than_v2 = false;
+      auto deps1 = getSubexprDependency(v1, variables_set);
+      auto deps2 = getSubexprDependency(v2, variables_set);
+      if (hasTensor(deps2)) {
+        return true;
+      }
+      if (hasTensor(deps1)) {
+        return false;
+      }
+      for (auto v : variables) {
+        if (deps1.count(v.variable) > 0 && deps2.count(v.variable) == 0) {
+          v1_lefter_than_v2 = false;
+        } else if (
+            deps2.count(v.variable) > 0 && deps1.count(v.variable) == 0) {
+          v1_lefter_than_v2 = true;
+        }
+      }
+      return v1_lefter_than_v2;
+    };
+    std::sort(sorted_inputs.begin(), sorted_inputs.end(), compare);
+    return sorted_inputs;
+  }
+
+ protected:
+  // Add a new value as an input of this expression. If `value` is a constant
+  // scalar, then evalate this constant and merge it with the constant term.
+  // Otherwise, add it as an input.
+  void addInput(Val* value) {
+    TORCH_CHECK(
+        hasSimilarType(dtype(), *value->getDataType()),
+        "Input types should be similar, but got: ",
+        dtype(),
+        ", and ",
+        *value->getDataType());
+    value = foldConstants(value);
+    if (hasValue(value)) {
```

Comment:
I don't get why `hasValue` isn't just calling `isConstScalar` since you're checking inputs not all the leaves.

Reply:
Now I get it. When you fold constants you may create a new scalar that has a single constant value, rather than checking what's in value can be evaluated to a constant. But can't we still use `isConstScalar` for this check? Seems like `isConstScalar` should have the `hasValue` shortcut in it.

Reply:
Oh sorry, didn't see your reply here when posting https://github.com/csarofeen/pytorch/pull/2215#discussion_r1046393717. I think we should use `isConst` instead of `isConstScalar`. I think the reason not using `isConstScalar` is because there are cases where `isConstScalar` is true, but expr evaluator can not evaluate it, for example for `ComplexDouble`. If this is the case, I don't want to treat it like a constant here.

Reply:
changed to `isConst`